### PR TITLE
Issue #548, #553 - CASE works with SELECT

### PR DESCRIPTION
**Pull request by seojungmin**
This patch adds a fix on the current parser.  The previous test scenario in #548 is tested and I added some scenarios including those mentioned in #553.

### CODE REVIEW

**Comment by pmenon** [comment]
This doesn't look right. 

1. It's a bad idea to use/throw exceptions to indicate conditions in expression/operator logic.
2. Why are two separate classes needed?
3. Why aren't smart pointers being used?

The design of CASE from [here](https://github.com/cmu-db/peloton/pull/119) is the correct approach. Use it as a starting point.

**Comment by seojungmin** [comment]
I was unaware that this code part that I ported has been refactored once.  The refactored design you mentioned could be applied. Thanks.

**Comment by apavlo** [comment]
What is the status of this PR?

**Comment by apavlo** [comment]
@seojungmin Why did the coverage go down so much?

**Comment by seojungmin** [comment]
@apavlo I am trying to see whether the coverage may differ if I rebase.

File: src/include/expression/case_expression.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    // Otherwise, we assume that the When Clauses provided are sufficient
+    if (argument_ != nullptr) {
+      for (uint32_t i = 0; i < clauses_.size(); i++)
+		clauses_[i].first.reset(
+           new peloton::expression::ComparisonExpression(
+              peloton::ExpressionType::COMPARE_EQUAL, argument_->Copy(),
+              clauses_[i].first->Copy()));
+    }
+  }
+  type::Value Evaluate(const AbstractTuple *tuple1, const AbstractTuple *tuple2,
```
Comment (pmenon) [inline src/include/expression/case_expression.h]: I think we should have two constructors: one with an argument and one without an argument. This makes it's use a little clearer. Thoughts?

File: test/expression/expression_test.cpp [hunk 2]
```diff
@@ -245,5 +246,188 @@ TEST_F(ExpressionTests, ExtractDateTests) {
...
+  tuple->SetValue(0, type::ValueFactory::GetIntegerValue(2), nullptr);
+  tuple->SetValue(1, type::ValueFactory::GetIntegerValue(1), nullptr);
+  result = case_expression->Evaluate(tuple, nullptr, nullptr);
+  expected = type::ValueFactory::GetIntegerValue(3);
+  EXPECT_EQ(type::CmpBool::CMP_TRUE, expected.CompareEquals(result));
+
+  delete case_expression;
+  delete schema;
+  delete tuple;
+}
```
Comment (pmenon) [inline test/expression/expression_test.cpp]: Can you modify these tests to use smart pointers?

**Comment by pmenon** [comment]
Overall, it looks good 👍 Just address the reviews, but otherwise LGTM.

File: src/include/expression/case_expression.h (hunk 1 above)
Comment (seojungmin) [inline src/include/expression/case_expression.h]: Agreed.  I am a natural born C programmer, and I hate introducing functions. BUT, I will take your suggestion.

**Comment by seojungmin** [comment]
@apavlo Can you merge this? I think it is ready.

**Pull request merged**